platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/linking/AbstractInOrderLinker.java

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (C) 2022-2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.event.linking;
+
+import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
+
+import com.swirlds.common.context.PlatformContext;
+import com.swirlds.common.crypto.Hash;
+import com.swirlds.common.sequence.map.SequenceMap;
+import com.swirlds.common.sequence.map.StandardSequenceMap;
+import com.swirlds.common.utility.throttle.RateLimitedLogger;
+import com.swirlds.platform.EventStrings;
+import com.swirlds.platform.consensus.EventWindow;
+import com.swirlds.platform.event.AncientMode;
+import com.swirlds.platform.event.EventCounter;
+import com.swirlds.platform.event.GossipEvent;
+import com.swirlds.platform.eventhandling.EventConfig;
+import com.swirlds.platform.gossip.shadowgraph.Shadowgraph;
+import com.swirlds.platform.internal.EventImpl;
+import com.swirlds.platform.system.events.BaseEventHashedData;
+import com.swirlds.platform.system.events.EventDescriptor;
+import com.swirlds.platform.wiring.NoInput;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Links events to their parents. Expects events to be provided in topological order.
+ * <p>
+ * Will not link events to parents in the following cases:
+ * <ul>
+ *     <li>The parent is ancient</li>
+ *     <li>The parent's generation does not match the generation claimed by the child event</li>
+ *     <li>The parent's time created is greater than or equal to the child's time created</li>
+ * </ul>
+ * Note: This class doesn't have a direct dependency on the {@link Shadowgraph ShadowGraph},
```

Comment:
This comment is no longer strictly true for this base class. Perhaps it should be moved down into the `GossipLinker`

Reply:
moved